rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,158 @@
+kubeconfig generation script to not create those cluster role bindings, but it will be enabled by default. 
+
+### Automatically provisioning RBAC resources to Kubernetes clusters.
+
+We will add a new type of resources, called "KubeProvision", where user will be able to specify Kubernetes RBAC resources they want to
+be provisioned onto their Kubernetes clusters.
+clusters. A single Teleport resource of that type can define multiple Kubernetes RBAC resources. There's a limit on size of a resource Teleport
+can save on the backend, therefore, if some users would have unusually large amount of Kubernetes RBAC resources they want to provision, they would
+need to split it into two KubeProvision resources.
+
+```protobuf
+import "teleport/header/v1/metadata.proto";
+
+// KubeProvision represents a Kubernetes resources that can be provisioned on the Kubernetes clusters.
+// This includes roles/role bindings and cluster roles/cluster role bindings.
+// For rationale behind this type, see the RFD 174.
+message KubeProvision {
+  // The kind of resource represented.
+  string kind = 1;
+  // Not populated for this resource type.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
+  // Common metadata that all resources share.
+  teleport.header.v1.Metadata metadata = 4;
+  // The specific properties of kube provision.
+  KubeProvisionSpec spec = 5;
+}
+
+// KubeProvisionSpec is the spec for the kube provision message.
+message KubeProvisionSpec {
+  // resources_data is base64 encoded YAML definitions of the Kubernetes resources.
+  string resources_data = 3;
+}
+```
+
+Every five minutes we will run a reconciliation loop that will compare the current state on Kubernetes clusters under
+Teleport management with the desired state and update the cluster's RBAC if there's a difference.
+
+Teleport will mark RBAC resources under its control with the "app.kubernetes.io/managed-by: Teleport" label. That way we will
+separate resources managed by Teleport and those managed by the user manually.
+
+Resource labels will be taken into account when doing the reconciliation - users will be able to match different
+Kubernetes clusters for different KubeProvision resources. If a KubeProvision resource doesn't have any labels defined it
+will not match any clusters, effectively being disabled.
+
+In order to be able to reconcile the state, Teleport will need to perform CRUD operations on Kubernetes RBAC resources.
+When performing CRUD operations, Teleport will impersonate the "default-cluster-admin" group, that was described in the previous
+section. This means that only newly enrolled
+Kubernetes clusters, or cluster where the user performed a manual upgrade of permissions will be in scope of provisioning the resources,
+since otherwise the "default-cluster-admin" role won't be available.
+
+## Security
+
+The introduction of this functionality does not directly impact the security of Teleport
```

Comment:
the kube agent already has abitilty to impersonate all groups/users, so they can always impersonate someone in the cluster